### DISCUSSION TITLE
Where to put transition animation code?

### DISCUSSION THREAD

**Discussion by fnick851** [statelyai/xstate#3168] [Q&A] [answered]
Hi,

Say I have a simple state machine: circleState -> EVENT -> squareState

When the EVENT happens, I want to play a morph animation morphing the circle into a square, then change to the squareState.
How should I model this? Where should I put the animation code? Say I am using GSAP to change a SVG circle to a square.

Should the animation itself be its own state? e.g. if the animation is 2s, then circleState -> EVENT -> morphAnimation(2s) -> ANIMATION_END_EVENT -> squareState. If this is the correct model, how do I change to the next state right after the animation finishes?

Thanks.

**Comment by davidkpiano** [comment, marked as answer]
It doesn't necessarily have to be. You can model the actual animation using child states:
![CleanShot 2022-03-23 at 08 19 32](https://user-images.githubusercontent.com/1093738/159697751-4bde594e-8a19-4a2a-9996-51cd96797d09.png)

The reason I like modeling it this way is because instead of thinking "there's a circle, now there's a circle changing into a square, now there's a square", you think of it more as a storyboard:

1. There's a circle
2. Now there's a square (animating from a circle until it's finished)

**Reply by fnick851** [reply]
That looks cleaner. Thanks!

**Reply by leevigraham** [reply]
@davidkpiano Would you use the same logic for a slide out drawer:

![Screenshot 2022-11-14 at 2 00 19 pm](https://user-images.githubusercontent.com/25124/201566027-73450dc9-7e6f-4e9b-8b81-38592fcbebca.png)

**Reply by davidkpiano** [reply]
Yes I would 👍

**Reply by leevigraham** [reply]
If the initial state is `closed` would this trigger the `closed.animating` state? Any way to avoid that.

Also could the `closed.animating` state be changed to `closed.closing`?

![image](https://user-images.githubusercontent.com/25124/201764023-3617a3b6-6dd7-42b4-8ed8-35b53b7f86da.png)

**Reply by davidkpiano** [reply]
Then another strategy is to have the default states be the "finished" states, and when transitioning, transition directly to the animating states.
![CleanShot 2022-11-14 at 14 13 50@2x](https://user-images.githubusercontent.com/1093738/201766889-02c18851-4ba7-483e-bd80-f9c6da899629.png)

**Comment by tsxoxo** [comment]
Modelling some animations here and this is gold: the ideas, the screenshots.

Thank You all for asking and answering!
